File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.cpp
class Solution {
public:
    int findMin(vector<int> &nums) {
        auto l{0};
        auto r{static_cast<int>(nums.size()) - 1};
        while (l < r) {
            auto m{l + (r - l) / 2};
            if (nums[l] < nums[m] && nums[m] < nums[r]) {
                // whole thing must be sorted
                return nums[l];
            } else if (l == m) {
                return min(nums[l], nums[r]);
            } else if (nums[l] < nums[m]) {
                l = m + 1;
            } else {
                r = m;
            }
        }
        return nums[r];
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [0] * len(nums)
        i = 0
        while i < len(nums):
            if i == 0:
                res[i] = 1
            else:
                res[i] = res[i - 1] * nums[i - 1]
            i += 1
        
        print(res)
        prev_post = 1
        i = len(nums) - 1
        while i >= 0:
            if i != len(nums) - 1:
                prev_post *= nums[i + 1]
                res[i] *= prev_post
            
            i -= 1 
        
        return res